### ISSUE TITLE
Asserting dictionary values with EqualTo().Using() fails if the value type is string

### ISSUE THREAD

**Issue by EvrenJG** [nunit/nunit#3127]
In more details, asserting with EqualTo().Using() that an IDictionary<TKey, TValueActual> by comparing to an IDictionary<TKey, TValueExpected>, where TValueActual and TValueExpected are different types and one is string seems to not function as designed.
For a simplified version of my problem, let's say that I have an actual value:
`var actual = new Dictionary<string, int>{ ["Key1"] = 3};`

I want to assert its values using another dictionary:
`var expected = new Dictionary<string, string>{["Key1"] = "3"};`

There are two ways I would have expected to be able to do that, and those two ways have two different issues preventing it. I could probably have opened two tickets, but maybe one or both of these behaviors are by design and really there's another solution to this.

Both ways use the `NUnitEqualityComparer.AreEqual` behavior, where when comparing two dictionaries, we first look for an external comparer (as defined in the various `Using` method overloads) accepting the dictionaries, and if not found fallback to the standard `DictionariesComparer`, which then looks into the same set of external comparers for one that accepts the values.

**1. Using a predicate**

```
Assert.That(
    actual,
    Is
        .EqualTo(expected)
        .Using<int, string>((actualValue, expectedValue) => actualValue.ToString() == expectedValue));
```

The problem with this is that the code assumes that my predicate relates to the dictionaries being compared, and not the values. This is due to the fact that the method `PredicateEqualityAdapter<TActual, TExpected>.CanCompare` returns always true.
This is strange because one of the advantages of having generic parameters would have been that it would allow writing CanCompare as `x is TActual and y is TExpected`. This in turn would have allowed the `NUnitEqualityComparer.AreEqual` method to correctly ignore the predicate for the dictionary comparison and then detect it for the value comparison!
Any reason why this method returns always true?

**2. Using `IEqualityComparer` (also same thing happens with `IComparer`)**

```
Assert.That(actual, Is.EqualTo(expected).Using(new StringEqualityComparer());

private class StringEqualityComparer : IEqualityComparer
{
    bool IEqualityComparer.Equals(object expected, object actual) => Equals(actual.ToString(), expected);

    public int GetHashCode(object obj) => throw new NotImplementedException();
}
```

The issue with this lies in `EqualityAdapter.CanCompare`:
```
public virtual bool CanCompare(object x, object y)
{
    if (x is string && y is string)
        return true;

    if (x is IEnumerable || y is IEnumerable)
        return false;

    return true;
}
```
Now this is called twice. The first time to see if it can be used to compare the dictionaries. It correctly returns false. The second time to see if it can be used to compare the dictionary values. It also returns false, but I think it should return false. The reason is that _string is an IEnumerable_, and the expected dictionary values are strings... I think this code was written expecting that x and y are of the same type (which is true unless someone uses external comparers in this way). So this would work if instead of this code it was:
```
public virtual bool CanCompare(object x, object y)
{
    if (x is string && y is string)
        return true;

    if (x is IEnumerable && !(x is string) || y is IEnumerable && !(y is string))
        return false;

    return true;
}
```
Basically just refusing to compare an enumerable to anything, unless that enumerable is a string.
If you like compact statements, this entire method can actually be simplified as:
`return (x is string || !(x is IEnumerable)) && (y is string || !(y is IEnumerable));`

**Comment by EvrenJG** [comment]
Here's a test class to reproduce both issues:

```
using System;
using System.Collections;
using System.Collections.Generic;
using NUnit.Framework;

namespace Tests
{
    public class Tests
    {
        [Test]
        public void Test1()
        {
            Assert.That(
                new Dictionary<string, int> {["1"] = 3},
                Is
                    .EqualTo(new Dictionary<string, string> {["1"] = "3"})
                    .Using<int, string>((actual, expected) => actual.ToString() == expected));
        }

        [Test]
        public void Test2()
        {
            Assert.That(
                new Dictionary<string, int> {["1"] = 3},
                Is
                    .EqualTo(new Dictionary<string, string> {["1"] = "3"})
                    .Using(new StringEqualityComparer()));
        }

        private class StringEqualityComparer : IEqualityComparer
        {
            bool IEqualityComparer.Equals(object expected, object actual) => Equals(actual.ToString(), expected);
        
            public int GetHashCode(object obj) => throw new NotImplementedException();
        }
    }
}
```

**Comment by EugeneKrapivin** [comment]
Hey!
From my understanding (I maybe wrong though) `EqualTo` compares the given objects, hence when using it in the comparing lambda you pass in `Test1` and the `StringEqualityComparer` you pass in `Test2` the tests try to cast the `Dictionary<TKey,TValue>` to a `string` or `int` which throws an understandable casting exception.

```
Failed   Test1
Error Message:
 System.InvalidCastException : Unable to cast object of type 'System.Collections.Generic.Dictionary`2[System.String,System.Int32]' to type 'System.Int32'.
```

Since `Dictionary<TKey, TValue>` is a collection I'd use collection `EquivalentTo` constraint. However, this doesn't solve all the problems because this method will try to check for equivalency between the `KeyValuePair<TKey, TValue>`s of each dictionary. To handle this you could use the `Using` comparison predicate like this:

```C#
[Test]
public void Test3()
{
    var actualDictionary = new Dictionary<string, int> {["1"] = 3};
    var expectedDictionary = new Dictionary<string, string> {["1"] = "3"};
    
    Assert.That(actualDictionary, Is.EquivalentTo(expectedDictionary)
        .Using<KeyValuePair<string, int>, KeyValuePair<string,string>>((actual, expected) 
            => actual.Key.Equals(expected.Key) && actual.Value.ToString().Equals(expected.Value)));
}
```

**Comment by stevenaw** [comment]
Thanks for providing your workaround @EugeneKrapivin 
I'm going to close this as there hasn't been any activity for a few years. Another way to approach this on newer runtimes now could also be to use an `IAlternateEqualityComparer`.
